File: clojure/scrabble-score/src/scrabble_score.clj
(ns scrabble-score
  (:require [clojure.string :as str]))

(def letters-score
  {1 "AEIOULNRST"
   2 "DG"
   3 "BCMP"
   4 "FHVWY"
   5 "K"
   8 "JX"
   10 "QZ"})

(defn- score-of-letter [c score letters]
  (if (>= (.indexOf letters (str/upper-case (str c))) 0) score 0))

(defn score-letter [c]
  (reduce + (for [[score letters] letters-score] (score-of-letter c score letters))))

(defn score-word [word]
  (reduce + (for [[score letters] letters-score
                  c word] (score-of-letter c score letters))))

File: javascript/scrabble-score/scrabble-score.js
var scoreTable = {
    AEIOULNRST: 1,
    DG: 2,
    BCMP: 3,
    FHVWY: 4,
    K: 5,
    JX: 8,
    QZ: 10
};

function scoreOf (element) {
    'use strict';

    var keys = Object.keys(scoreTable);
    for (var i = 0; i < keys.length; i ++) {
        if (keys[i].indexOf(element) !== -1) {
            return scoreTable[keys[i]];
        }
    }
    return 0;
}

function score (txt) {
    'use strict';

    if (!txt) {
        return 0;
    }
    var chars = txt.split('').map(function (element) {
        return element.toUpperCase();
    });
    return chars.reduce(function (total, element) {
        return total += scoreOf(element);
    }, 0);
}

module.exports = score;

